src/Planner/CollectTableExpressionData.cpp

```diff
@@ -102,45 +114,51 @@ class CollectSourceColumnsVisitor : public InDepthQueryTreeVisitor<CollectSource
-        table_expression_data.addColumn(column_node->getColumn(), column_identifier);
+        auto column_identifier = planner_context->getGlobalPlannerContext()->createColumnIdentifier(node);
+        table_expression_data.addColumn(column_node->getColumn(), column_identifier, select_added_columns);
     }
 
-    static bool needChildVisit(const QueryTreeNodePtr & parent, const QueryTreeNodePtr & child_node)
+    static bool isAliasColumn(const QueryTreeNodePtr & node)
     {
-        if (auto * join_node = parent->as<JoinNode>())
-        {
-            if (join_node->getJoinExpression() == child_node && join_node->isUsingJoinExpression())
-                return false;
-        }
-        auto child_node_type = child_node->getNodeType();
-        return !(child_node_type == QueryTreeNodeType::QUERY || child_node_type == QueryTreeNodeType::UNION);
+        const auto * column_node = node->as<ColumnNode>();
+        if (!column_node || !column_node->hasExpression())
+            return false;
+        const auto & column_source = column_node->getColumnSourceOrNull();
+        if (!column_source)
+            return false;
+        return column_source->getNodeType() != QueryTreeNodeType::JOIN &&
+               column_source->getNodeType() != QueryTreeNodeType::ARRAY_JOIN;
     }
 
-private:
-
-    void visitUsingColumn(QueryTreeNodePtr & node)
+    static bool needChildVisit(const QueryTreeNodePtr & parent_node, const QueryTreeNodePtr & child_node)
     {
-        auto & column_node = node->as<ColumnNode&>();
-        if (column_node.hasExpression())
-        {
-            auto & table_expression_data = planner_context.getOrCreateTableExpressionData(column_node.getColumnSource());
-            bool column_already_exists = table_expression_data.hasColumn(column_node.getColumnName());
-            if (column_already_exists)
-                return;
-
-            auto column_identifier = planner_context.getGlobalPlannerContext()->createColumnIdentifier(node);
-            table_expression_data.addAliasColumnName(column_node.getColumnName(), column_identifier);
+        auto child_node_type = child_node->getNodeType();
+        return !(child_node_type == QueryTreeNodeType::QUERY ||
+                 child_node_type == QueryTreeNodeType::UNION ||
+                 isAliasColumn(parent_node));
+    }
 
-            visitImpl(column_node.getExpressionOrThrow());
-        }
-        else
-            visitImpl(node);
+    void setKeepAliasColumns(bool keep_alias_columns_)
+    {
+        keep_alias_columns = keep_alias_columns_;
     }
 
-    PlannerContext & planner_context;
+private:
+    PlannerContextPtr & planner_context;
+
+    /// Replace ALIAS columns with their expressions or register them in table expression data.
+    /// Usually we can replace them when we build some "local" actions DAG
+    /// (for example Row Policy or PREWHERE) that is applied on right top of table expression.
+    bool keep_alias_columns = true;
```

Comment:
When can not we do it?

Reply:
In other cases, when we keep ColumnNode with expression and compute after `read`. Added it to comment